src/Command/RenderBlockCommand.php

```diff
@@ -41,21 +87,22 @@ public function configure(): void
         );
 
         $this->addArgument('manager', InputArgument::REQUIRED, 'The manager service id');
-        $this->addArgument('page_id', InputArgument::REQUIRED, 'The page id');
-        $this->addArgument('block_id', InputArgument::REQUIRED, 'The page id');
+        $this->addArgument('block_id', InputArgument::REQUIRED, 'The block id');
     }
 
     protected function execute(InputInterface $input, OutputInterface $output): int
     {
-        $manager = $this->getContainer()->get($input->getArgument('manager'));
+        $manager = $input->getArgument('manager');
 
-        if (!$manager instanceof CmsManagerInterface) {
-            throw new \RuntimeException('The service does not implement the CmsManagerInterface');
+        if (!\in_array($manager, ['sonata.page.cms.snapshot', 'sonata.page.cms.page'], true)) {
+            throw new \RuntimeException(
+                'Available managers are "sonata.page.cms.snapshot" and "sonata.page.cms.page"'
+            );
         }
 
-        $page = $manager->getPageById($input->getArgument('page_id'));
+        $managerService = 'sonata.page.cms.snapshot' === $manager ? $this->cmsSnapshotManager : $this->cmsPageManager;
```

Comment:
You're right, I did not think about it. I'm trying to inject a service locator containing all the services implementing `CmsManagerInterface` as you suggest and I let you know.

Reply:
Hello @VincentLanglet, I tried, as you suggest, to add a service locator for cms manager services but I'm struggling while trying to add services as argument of the service locator in a compiler pass. See last commit of the PR: https://github.com/sonata-project/SonataPageBundle/pull/1422/commits/59edbd5b067a2694565f2d14d7ce64098e8f83ca.

I have the following error:
![error_during_compiler_pass_execution](https://user-images.githubusercontent.com/2642302/173199111-6cf34600-9284-4a94-ad61-7b9a58739a47.png). I guess using reflection in a compiler pass is not a good idea, any other idea how to add services implementing [this interface](https://github.com/sonata-project/SonataPageBundle/blob/f56de820e5d14c008238dbf7bcb077a4c963f717/src/CmsManager/CmsManagerInterface.php), maybe using tags? But it should be documented somewhere. What do you think?

Reply:
I know @franmomu already used service locator multiple times, maybe he can help.

If you look at https://github.com/sonata-project/SonataAdminBundle/pull/7056/files, reflection was used so this might not be the origin of the problem.

I think the purpose of the compilerPass is to dynamically call `$definition->replaceArgument` in order to set the dependency. I don't think this is what you're doing here.

Reply:
Yes but I think this is failing in my case because I don't loop on services with a specific tag but onto all the services. And I'm using reflection on a service that points to a class that has code outside of the class scope, like [this one](https://github.com/symfony/cache/blob/5.4/Psr16Cache.php#L22-L24). So the code is executed and my guess is that the file is not autoloaded yet so it is failing.

I think I should use `addArgument()` as I did and not `replaceArgument()` because I guess we want to support both "natives" services that implements the [CmsManagerInterface](https://github.com/sonata-project/SonataPageBundle/blob/f56de820e5d14c008238dbf7bcb077a4c963f717/src/CmsManager/CmsManagerInterface.php) and services that implements the same interface but defined by the user of SonataPageBundle.

Reply:
I don't this bundle well, but reading the comments we could maybe autoconfigure services that implement `CmsManagerInterface` with a tag, that would allow to loop them instead of all services.